File: easy/1331_rank_transform_of_an_array/solution.cpp
#include <set>
#include <vector>

class Solution {
public:
  std::vector<int> arrayRankTransform(std::vector<int> &arr) {
    std::set<int> arrSet(arr.begin(), arr.end());
    std::unordered_map<int, int> rankMap;

    int rank = 0;
    for (const int &num : arrSet)
      rankMap[num] = ++rank;

    for (size_t i = 0; i < arr.size(); ++i)
      arr[i] = rankMap[arr[i]];

    return arr;
  }
};

File: easy/1331_rank_transform_of_an_array/Solution.java
import java.util.HashMap;
import java.util.Map;
import java.util.SortedSet;
import java.util.TreeSet;

class Solution {
  public int[] arrayRankTransform(int[] arr) {
    final SortedSet<Integer> arrSet = new TreeSet<>();
    for (final int num : arr) {
      arrSet.add(num);
    }

    final Map<Integer, Integer> rankMap = new HashMap<>();

    int rank = 0;
    for (final int num : arrSet) {
      rankMap.put(num, rank);
      ++rank;
    }

    for (int i = 0; i < arr.length; ++i) {
      arr[i] = rankMap.get(arr[i]);
    }

    return arr;
  }
}
